### PR TITLE
auto parallel add dp demo.

### PR DESCRIPTION
**Pull request by winter-wang**
### PR Category

Auto Parallel

### PR Types

Not User Facing 
### Description

add dp precision verify unit test.

### Other
Pcard-67164

### CODE REVIEW

File: test/auto_parallel/pir/test_semi_auto_parallel_dist_to_static_pir.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def test_mlp(self):
             {"dtype": "float32", "seed": "2023"}, {"backend": ["gpu"]}
         )
         for envs in envs_list:
-            # self._log_dir.name = "./log"
+            self._log_dir.name = "./log"
```
Comment (JZ-LIANG) [inline test/auto_parallel/pir/test_semi_auto_parallel_dist_to_static_pir.py]: remember to remove this.

File: python/paddle/distributed/auto_parallel/static/reshard_funcs/s_to_r_reshard_func.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        padding_num=0,
+    ):
+        src_mesh = src_dist_attr.process_mesh
+        num_of_process = len(src_mesh.process_ids)
+        dtype = src_value.dtype
+        group = new_process_group(src_mesh.process_ids)
+        allgather_value = paddle._pir_ops.c_allgather(
+            src_value, group.id, num_of_process, False
+        )
+        allgather_value.set_type(dst_type)
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/static/reshard_funcs/s_to_r_reshard_func.py:95]: should think about a better method to remove this hack, like communication operation in pir using special infermeta_local

File: test/auto_parallel/pir/test_semi_auto_parallel_dist_to_static_pir.py (hunk 1 above)
Comment (winter-wang) [reply test/auto_parallel/pir/test_semi_auto_parallel_dist_to_static_pir.py]: done, thanks!

File: python/paddle/distributed/auto_parallel/static/reshard_funcs/s_to_r_reshard_func.py (hunk 2 above)
Comment (winter-wang) [reply python/paddle/distributed/auto_parallel/static/reshard_funcs/s_to_r_reshard_func.py:95]: as discussed, I will systematically solve this problem in the next pr.

**Review by XieYunshen — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]

**Review by sunzhongkai588 — APPROVED** [review]
LGTM

**Pull request merged**